### PR TITLE
Implement type safety for TLS

### PR DESCRIPTION
**Pull request by tamird**
I came across these type conversions while investigating https://bugs.fuchsia.dev/p/fuchsia/issues/detail?id=68325. It turns out that changing TLS from pthread to stdcpp (#26936) *did* resolve the UB on Fuchsia.

@ctiller @brunowonka

### CODE REVIEW

**Comment by tamird** [comment]
Oops, had some incorrect formatting. Craig, could I get another try run?

**Comment by tamird** [comment]
Some significant changes here to reduce the tls.h API surface. One more kokoro run please, Craig.

File: src/core/lib/iomgr/timer_generic.cc
Comment (tamird) [inline src/core/lib/iomgr/timer_generic.cc:223]: the comment above this block is slightly out of date now; let me know if you'd like me to update it. The incorrect part is "intptr_t" -- the real limitation is that TLS types must not be larger than a machine word.

File: src/core/lib/gpr/tls_stdcpp.h
Comment (ctiller) [inline src/core/lib/gpr/tls_stdcpp.h:27]: could we make this `static_assert(sizeof(type) == sizeof(intptr_t)); thread_local type`?

I'd like the config that I edit code with to have the same restrictions as the most restrictive platform we build for (otherwise my life is waiting for CI to come back for minor edits)

File: src/core/lib/iomgr/timer_generic.cc
Comment (ctiller) [reply src/core/lib/iomgr/timer_generic.cc:223]: I don't think it matters terribly - we'll nuke this code in a few months.

File: src/core/lib/gpr/tls_stdcpp.h
Comment (tamird) [reply src/core/lib/gpr/tls_stdcpp.h:27]: Good idea!

Prepending a static assertion doesn't quite work because then `static GPR_THREAD_LOCAL(std::string) greeting` expands to `static static_assert(...); thread_local std::string greeting` which doesn't compile. Moving the static assertion to the end of the expression will produce something that compiles today but will prevent initializers.

I might have to "pull in" the static modifier into the macro to get this working. Any ideas I haven't thought of?

**Comment by ctiller** [comment]
```
template <typename T, typename Ignored>
struct Wrapper;

template <typename T>
struct Wrapper<T, absl::enable_if_t<sizeof(T) < sizeof(intptr_t)>> {
  using Type = T;
};

#define GPR_THREAD_LOCAL(type) thread_local Wrapper<type>
```

**Comment by ctiller** [comment]
maybe better:

template <typename T>
struct Wrapper {
  static_assert(sizeof(T) <= sizeof(intptr_t));
  using Type = T;
};

**Comment by tamird** [comment]
Nice! Done, and made it apply uniformly to all toolchains (not just stdcpp).

File: src/core/lib/gpr/tls.h
Comment (ctiller) [inline src/core/lib/gpr/tls.h:30]: I know I typed Wrapper in the comment thread, but let's give it a more supportable name long term...

```
namespace grpc_core {
template <...> class TlsTypeWrapper { ... };
```

Comment (tamird) [reply src/core/lib/gpr/tls.h]: Sure, done. Renamed the pthread type to look more C++ish as well.

**Comment by tamird** [comment]
Collapsed all the implementations into the same header to avoid having to include tls.h from each one. Otherwise clang-tidy is unhappy:
```
Error while processing /var/local/git/grpc/src/core/lib/gpr/tls_pthread.h.
src/core/lib/gpr/tls_pthread.h:33:27: error: no template named 'TlsTypeWrapper' [clang-diagnostic-error]
class pthreadTlsWrapper : TlsTypeWrapper<T> {
                          ^
src/core/lib/gpr/tls_pthread.h:39:53: error: use of undeclared identifier 'key_' [clang-diagnostic-error]
  void Init() { GPR_ASSERT(0 == pthread_key_create(&key_, nullptr)); }
                                                    ^
src/core/lib/gpr/tls_pthread.h:42:40: error: use of undeclared identifier 'key_' [clang-diagnostic-error]
    GPR_ASSERT(0 == pthread_key_delete(key_));
                                       ^
src/core/lib/gpr/tls_pthread.h:43:5: error: use of undeclared identifier 'key_' [clang-diagnostic-error]
    key_ = 0;
```

**Comment by tamird** [comment]
Damn it, compilation failures on Apple systems:
```
Mon Aug  9 13:57:28 PDT 2021 - In file included from /Volumes/BuildData/tmpfs/src/github/grpc/workspace_objc_macos_opt_native/src/core/lib/iomgr/exec_ctx.h:32:
Mon Aug  9 13:57:28 PDT 2021 - /Volumes/BuildData/tmpfs/src/github/grpc/workspace_objc_macos_opt_native/src/core/lib/gpr/tls.h:123:24: error: no viable conversion from 'int' to 'pthread_attr_t' (aka '_opaque_pthread_attr_t')
Mon Aug  9 13:57:28 PDT 2021 -   pthread_key_t key_ = 0;
```
Removed initialization of pthread_key_t. Another run, please?

**Comment by tamird** [comment]
More compilation failures, I apparently forgot to include headers needed for pthread when I combined the headers. Should be fixed now.

**Comment by tamird** [comment]
Fixed formatting.

**Comment by ctiller** [comment]
This is good stuff... Will give it one more thorough review after breakfast.

One thing to note/discuss @veblush @nicolasnoble @markdroth -- #26912 passed the portability suites without using the GPR TLS stuff (I'd forgotten about the facility when I drafted the code) -- which warrants the question, do we know that we still need this facility, or would it suffice to just use the C++11 mechanism for supported compilers?

When I built this mechanism we were still very much in C land where the ecosystem here is substantially more fractured.

**Comment by markdroth** [comment]
Assuming that we can use the C++ interface inside of core, that seems like a better option than this GPR API.  The only reason I can think of why we might need the GPR API is if it's used by wrapped languages, so I'd suggest checking that.

**Comment by tamird** [comment]
I guess there was a reason at one time: https://github.com/grpc/grpc/commit/94c6766bf11f8d9e835988b946f178d3bac368b8

@veblush

**Comment by ctiller** [comment]
Yeah I guess I remembered the iOS thing... was hoping that was old enough to be gone now (is a 2016 phone OS old enough for us to drop support for?)

**Comment by tamird** [comment]
Per the description in #24572:

> Configuring port_platform to use `PTHREAD_TLS` instead of `STDCPP_TLS` when it's not available. This shouldn't happen to real devices but it can happen when running on the simulator. (e.g. iOS 9 32bits simulator)

So it isn't clear that this is really about phones. Anyway, removed all this stuff in #26960, let's see what the bots say there.

**Comment by veblush** [comment]
Our minimum supported iOS version is 9 so theoretically we should support iOS 9.0 on both 32 bits and 64 bits anyway. We bumped this version (https://github.com/grpc/grpc/pull/24282) so I don't think we can bump this again soon without strong reason. This affects all iOS applications relying on gRPC so it should be done carefully. So we should keep iOS 9 32 bits emulation support.

Also Google3 still depends on `GPR_PTHREAD_TLS` for both Android and iOS so we need to migrate them first. (I have CLs for them but they haven't passed the tests although I'm not sure those failures are valid)

**Comment by tamird** [comment]
Makes sense to me. I can send another PR after this one to retire GPR_GCC_TLS and GPR_MSVC_TLS if that sounds reasonable to folks.

**Comment by tamird** [comment]
Another question: should this move to `src/core/lib/gprpp/` since this is a C++ API now?

**Comment by markdroth** [comment]
If this is now a C++ API, I would be in favor of moving the C++ API to src/core/lib/gprpp.  If we still need the C99 API for wrapped languages, that part can be in src/core/lib/gpr, but it should just be a wrapper around the C++ API.

**Comment by tamird** [comment]
Wrapped languages do not use this API as far as I can tell. See these searches (all results are from this repo):
- https://github.com/search?l=&q=org%3Agrpc+GPR_TLS_DECL&type=code
- https://github.com/search?l=&q=org%3Agrpc+GPR_TLS_CLASS_DECL&type=code
- https://github.com/search?l=&q=org%3Agrpc+GPR_TLS_CLASS_DEF&type=code

Shall I go ahead and move it?

File: src/core/lib/gpr/tls.h
Comment (markdroth) [inline src/core/lib/gpr/tls.h:128]: It looks like there are a number of cases where we use a global thread-local variable.  Note that the style guide prohibits global variables that are not trivially destructible:

https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

Is the `pthreadTlsWrapper` class trivially destructible?  If not, this won't work.  If so, please add a static assertion that guarantees that property, with a comment as to why it's there.

File: src/core/lib/gpr/tls.h
Comment (markdroth) [inline src/core/lib/gpr/tls.h:130]: If we are making this a C++ API anyway, why not just change callers to directly use the C++ API, rather than wrapping it in this C-like interface?  We could presumably provide the same C++ API for all thread-local implementations.

Moving to gprpp makes sense for C++ code, but this C-style API would probably want to stay in gpr if we keep it.  But if there are no callers of this API outside of core, then we don't need the C-style API to begin with.

**Review by markdroth — CHANGES_REQUESTED** [review]

File: src/core/lib/gpr/tls.h
Comment (tamird) [reply src/core/lib/gpr/tls.h:130]: Because there's no wrapper class *being instantiated* in the not-pthread case. What would there be a method on?

Comment (tamird) [reply src/core/lib/gpr/tls.h:130]: Regarding the location: is that determined by the API *style* or by its language compatibility? This commit changes this header such that it won't compile in C, but I defer to you on what the heuristic should be.

File: src/core/lib/gpr/tls.h
Comment (tamird) [reply src/core/lib/gpr/tls.h]: Done.

File: src/core/lib/gpr/tls.h
Comment (tamird) [reply src/core/lib/gpr/tls.h]: Done.

File: src/core/lib/gpr/tls.h
Comment (markdroth) [reply src/core/lib/gpr/tls.h]: It's generally about the API style.  We can compile code written in C++ no matter what directory it's in.

If the API here is still a C-style API, then I guess keeping it in the gpr directory is fine.

File: src/core/lib/gprpp/tls.h
Comment (tamird) [reply src/core/lib/gprpp/tls.h:127]: This assertion is being applied to `PthreadTlsWrapper`, not `TlsTypeWrapper`.

Unlike `TlsTypeWrapper`, `PthreadTlsWrapper` *is* instantiated, so this assertion is necessary.

File: src/core/lib/gpr/tls.h
Comment (tamird) [reply src/core/lib/gpr/tls.h]: Ok, I will move it back. Anything else I should address?

Comment (tamird) [reply src/core/lib/gpr/tls.h:151]: Moved back to gpr.

File: src/core/lib/gprpp/tls.h
Comment (tamird) [reply src/core/lib/gprpp/tls.h]: Added comments to clarify what's what. Hopefully it's clearer now.

**Comment by tamird** [comment]
CI passed but formatting was off. Fixed now.

File: src/core/lib/gpr/tls.h
Comment (markdroth) [inline src/core/lib/gpr/tls.h:139]: Can this assertion be moved into the `PthreadTlsImpl` ctor, so that we don't need a separate class here?

Comment (tamird) [reply src/core/lib/gpr/tls.h:139]: The thing which must be trivially destructible is `PthreadtlsImpl<T>`, not `T` (though `T` *also* must be trivially destructible, but that's enforced separately by `TlsTypeConstrainer` and for a different reason)

**Review by ctiller — APPROVED** [review]

**Pull request merged**
